config/config.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	if len(s.Values) == 0 {
+		return 0, fmt.Errorf("empty system values in config")
+	}
+
+	// descriptorTableID = 3
+	// Search for the first key after the descriptor table.
+	// We can't use Get as we don't mind if there is nothing after
+	// the descriptor table.
+	key := keys.MakeTablePrefix(keys.DescriptorTableID + 1)
+	index := sort.Search(len(s.Values), func(i int) bool {
+		return bytes.Compare(s.Values[i].Key, key) >= 0
+	})
+
+	if index == 0 {
+		return 0, fmt.Errorf("descriptor table not found in system config of %d values", len(s.Values))
+	}
+
+	key = s.Values[index-1].Key
+
+	// Extract object ID from key.
+	// TODO(marc): move sql/keys.go to keys (or similar) and use a DecodeDescMetadataKey.
+	// We should also check proper encoding.
+	descriptorPrefix := keys.MakeTablePrefix(keys.DescriptorTableID)
+	remaining := bytes.TrimPrefix(key, descriptorPrefix)
+	// TrimPrefix returns the bytes unchanged if the prefix does not match.
+	if len(remaining) == len(key) {
+		return 0, fmt.Errorf("descriptor table not found in system config of %d values", len(s.Values))
+	}
+	// DescriptorTable.PrimaryIndex.ID
+	remaining, _ = encoding.DecodeUvarint(remaining)
+	// descID
+	_, id := encoding.DecodeUvarint(remaining)
+	return uint32(id), nil
+}
+
+// GetZoneConfigForKey looks up the zone config for the range containing 'key'.
+// It is the caller's responsibility to ensure that the range does not need to be split.
+func (s *SystemConfig) GetZoneConfigForKey(key proto.Key) (*ZoneConfig, error) {
+	if objectID, ok := ObjectIDForKey(key); ok {
+		return s.GetZoneConfigForID(objectID)
+	}
+	// Not in the structured data namespace.
+	return DefaultZoneConfig, nil
+}
+
+// GetZoneConfigForID looks up the zone config for the object (table or database)
+// with 'id'.
+func (s *SystemConfig) GetZoneConfigForID(id uint32) (*ZoneConfig, error) {
+	if ZoneConfigHook == nil {
+		return nil, util.Errorf("ZoneConfigHook not set, unable to lookup zone config")
+	}
+	// For now, only user databases and tables get custom zone configs.
+	if id <= keys.MaxReservedDescID {
+		return DefaultZoneConfig, nil
+	}
+	return ZoneConfigHook(s, id)
+}
+
+// ComputeSplitKeys takes a start and end key and returns an array of keys
+// at which to split the span [start, end).
+// The only required splits are at each user table prefix.
+func (s *SystemConfig) ComputeSplitKeys(startKey, endKey proto.Key) []proto.Key {
+	if TestingDisableTableSplits {
+		return nil
+	}
+
+	tableStart := proto.Key(keys.UserTableDataMin)
+	if !tableStart.Less(endKey) {
+		// This range is before the user tables span: no required splits.
+		return nil
+	}
+
+	startID, ok := ObjectIDForKey(startKey)
+	if !ok || startID <= keys.MaxReservedDescID {
+		// Start key is not in the structured data space, or part of the
+		// system space, use the first non-reserved ID.
+		startID = keys.MaxReservedDescID + 1
+	}
+
+	// Find the largest object ID.
+	endID, err := s.GetLargestObjectID()
```

Comment:
Ah, good point. You should comment as wasn't obvious. The common case is that `endKey` is going to point to a table-key. It might be worth optimizing for that case.